File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/extension/RepositoryLinkerExtensionFactory.java
package ru.tinkoff.kora.database.annotation.processor.extension;


import ru.tinkoff.kora.database.annotation.processor.DbUtils;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionFactory;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import java.util.Optional;

public class RepositoryLinkerExtensionFactory implements ExtensionFactory {
    @Override
    public Optional<KoraExtension> create(ProcessingEnvironment processingEnvironment) {
        var repository = processingEnvironment.getElementUtils().getTypeElement(DbUtils.REPOSITORY_ANNOTATION.canonicalName());
        if (repository == null) {
            return Optional.empty();
        } else {
            return Optional.of(new RepositoryKoraExtension(processingEnvironment));
        }
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/extension/RepositoryLinkerExtensionFactory.kt
package ru.tinkoff.kora.database.symbol.processor.extension

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import ru.tinkoff.kora.database.symbol.processor.DbUtils
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionFactory
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension

@KspExperimental
class RepositoryLinkerExtensionFactory : ExtensionFactory {
    override fun create(resolver: Resolver, kspLogger: KSPLogger, codeGenerator: CodeGenerator): KoraExtension? {
        val repository = resolver.getClassDeclarationByName(resolver.getKSNameFromString(DbUtils.repositoryAnnotation.canonicalName))
        return repository?.let { RepositoryKoraExtension(kspLogger) }
    }
}
